File: python/binary-tree.py
class Node:

  def __init__(self, data):
    self.data = data
    self.left = None
    self.right = None


class BinarySearchTree:

  def __init__(self):
    self.root = None

  def insert(self, data):
    new_node = Node(data)

    if self.root is None:
      self.root = new_node
    else:
      self.insert_node(self.root, new_node)

  def insert_node(self, node, new_node):
    if new_node.data < node.data:
      if node.left is None:
        node.left = new_node
      else:
        self.insert_node(node.left, new_node)
    else:
      if node.right is None:
        node.right = new_node
      else:
        self.insert_node(node.right, new_node)

  def remove(self, data):
    self.root = self.remove_node(self.root, data)

  def remove_node(self, node, key):
    if node is None:
      return None
    elif key < node.data :
      node.left = self.remove_node(node.left, key)
      return node
    elif key > node.data:
      node.right = self.remove_node(node.right, key)
      return node
    else:
      if node.left is None and node.right is None:
        node = None
        return node
      elif node.left is None:
        node = node.right
        return node
      elif node.right is None:
        node = node.left
        return node
      else:
        aux = self.find_min_node(self.right)
        node.data = aux.data

        node.right = self.remove_node(node.right, aux.data)
        return node

  # helper methods

  def find_min_node(self, node):
    if node.left is None:
      return node
    else:
      return self.find_min_node(node.left)

  def traverse_in_order(self, node):
    if node is not None:
      self.traverse_in_order(node.left)
      print(node.data)
      self.traverse_in_order(node.right)

  def get_root_node(self):
    return self.root

bst = BinarySearchTree()
bst.insert(15) 
bst.insert(25) 
bst.insert(10) 
bst.insert(7) 
bst.insert(22) 
bst.insert(17) 
bst.insert(13) 
bst.insert(5) 
bst.insert(9) 
bst.insert(27)

root = bst.get_root_node()
bst.traverse_in_order(root)

bst.remove(5)

root = bst.get_root_node()
bst.traverse_in_order(root)

File: javascript/binary-tree.js
class Node {
	constructor(data) {
		this.data = data;
		this.left = null;
		this.right = null;
	}
}

class BinarySearchTree {
	constructor() {
		this.root = null;
	}

	insert(data) {
		let newNode = new Node(data);

		if(this.root === null)
			this.root = newNode;
		else
			this.insertNode(this.root, newNode);
	}

	insertNode(node, newNode) {
		if(newNode.data < node.data) {
			if(node.left === null)
				node.left = newNode;
			else
				this.insertNode(node.left, newNode);
		} else {
			if(node.right === null)
				node.right = newNode;
			else
				this.insertNode(node.right, newNode);
		}
	}

	remove(data) {
		this.root = this.removeNode(this.root, data);
	}

	removeNode(node, key) {
		// if the root is null then tree is  
		// empty 
		if(node === null)
			return null;

		// if data to be deleted is less than  
		// roots data then move to left subtree 
		else if(key < node.data) {
			node.left = this.removeNode(node.left, key);
			return node;
		}

		// if data to be deleted is greater than  
		// roots data then move to right subtree 
		else if(key > node.data) {
			node.right = this.removeNOde(node.right, key);
			return node;
		}

		// if data is similar to the root's data  
		// then delete this node 
		else {
			// deleting node with no children
			if(node.left === null && node.right == null) {
				node = null;
				return node;
			}

			// deleting node with one child
			if(node.left === null) {
				node = node.right;
				return node;
			}

			else if(node.right === null) {
				node = node.left;
				return node;
			}

			// deleting node with two children
			// minimum node of the right subtree is stored in aux
			let aux = tihs.findMinNode(node.right);
			node.data = aux.data;

			node.right = this.removeNOde(node.right, aux.data);
			return node;
		}

	}

	// helper functions

	inorder(node) {
		if(node !== null) {
			this.inorder(node.left);
			console.log(node.data);
			this.inorder(node.right);
		}
	}

	preorder(node) {
		if(node !== null) {
			console.log(node.data);
			this.preorder(node.left);
			this.preorder(node.right);
		}
	}

	postorder(node) {
		if(node !== null) {
			this.postorder(node.left);
			this.postorder(node.right);
			console.log(node.data);
		}
	}

	findMinNode(node) {
		if(node.left === null)
			return node;
		else
			return this.findMinNode(node.left);
	}

	getRootNode() {
		return this.root;
	}

	search(node, data) {
		if(node === nulll)
			return null;

		else if(data < node.data)
			return this.search(node.left, data);

		else if(data > node.data)
			return this.search(node.right, data);

		else
			return node;
	}
}


module.exports = BinarySearchTree;
